### ISSUE TITLE
Range query for 3 lines in a 6 Nodes cluster takes about 1300 conversation iterations between the Coordinator and Replica nodes

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#2237]
*Installation details*
Scylla version (or git commit hash): b84828b48794a7e2e5281cc813ffb7a5b01e07cf
Cluster size: 6 (ccm cluster)

*Description*

1. `ccm create scylla-6 --scylla --vnodes -n 6 --install-dir=<scylla dir>`
2. `ccm start --jvm_arg="--memory" --jvm_arg="1G" --jvm_arg="--collectd-address" --jvm_arg="127.0.0.1:25826" --jvm_arg="--collectd" --jvm_arg="1" --jvm_arg="--default-log-level" --jvm_arg="info"`
3. From `cqlsh`:
```
CREATE KEYSPACE ks2   WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 };
CREATE TABLE ks2.test (   a int PRIMARY KEY,   b text );
INSERT INTO ks2.test (a, b) VALUES ( 3, '1' );
INSERT INTO ks2.test (a, b) VALUES ( 2, '2' );
INSERT INTO ks2.test (a, b) VALUES ( 1, '3' );
TRACING ON
SELECT * from ks2.test ;
TRACING OFF
SELECT * from system_traces.sessions ;
``` 
I received the following output:
```
fred@cqlsh> SELECT * from system_traces.sessions ;

 session_id                           | client    | command | coordinator | duration | parameters                                                                                                                                                        | request            | started_at
--------------------------------------+-----------+---------+-------------+----------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------+--------------------------
 5973c320-14e3-11e7-b84f-000000000000 | 127.0.0.1 |   QUERY |   127.0.0.1 |   536961 | {'consistency_level': 'ONE', 'page_size': '100', 'query': 'SELECT * from ks2.test ;', 'serial_consistency_level': 'SERIAL', 'user_timestamp': '1490835234898519'} | Execute CQL3 query | 2017-03-30 00:53:54+0000
```

Take the `session_id` value from the output above (in my case it's `5973c320-14e3-11e7-b84f-000000000000`) and use it for selecting the corresponding entries in the `system_traces.events` CF:
```
fred@cqlsh> SELECT COUNT(*) from system_traces.events where session_id=5973c320-14e3-11e7-b84f-000000000000;

 count
-------
  5192

(1 rows)
fred@cqlsh> SELECT * from system_traces.events where session_id=5973c320-14e3-11e7-b84f-000000000000;

 session_id                           | event_id                             | activity                                                              | parent_id       | source    | source_elapsed | span_id         | thread
--------------------------------------+--------------------------------------+-----------------------------------------------------------------------+-----------------+-----------+----------------+-----------------+---------
 5973c320-14e3-11e7-b84f-000000000000 | 5973e78c-14e3-11e7-b84f-000000000000 |                                                   Parsing a statement | 106811168079939 | 127.0.0.1 |              2 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5973e998-14e3-11e7-b84f-000000000000 |                                                Processing a statement | 106811168079939 | 127.0.0.1 |             54 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59742b71-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.5 | 106811168079939 | 127.0.0.1 |           1740 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597431f9-14e3-11e7-a72a-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.5 |             12 |  47600206259384 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597435d8-14e3-11e7-a72a-000000000000 | read_mutation_data handling is done, sending a response to /127.0.0.1 | 106811168079939 | 127.0.0.5 |            111 |  47600206259384 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5974403c-14e3-11e7-b84f-000000000000 |                      read_mutation_data: got response from /127.0.0.5 | 106811168079939 | 127.0.0.1 |           2272 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59744247-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.3 | 106811168079939 | 127.0.0.1 |           2324 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59744673-14e3-11e7-88ef-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.3 |             12 | 202113647559205 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59744a6a-14e3-11e7-88ef-000000000000 | read_mutation_data handling is done, sending a response to /127.0.0.1 | 106811168079939 | 127.0.0.3 |            114 | 202113647559205 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59745425-14e3-11e7-b84f-000000000000 |                      read_mutation_data: got response from /127.0.0.3 | 106811168079939 | 127.0.0.1 |           2782 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5974553f-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.4 | 106811168079939 | 127.0.0.1 |           2810 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59745995-14e3-11e7-bb16-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.4 |             10 | 212992572257781 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59745daa-14e3-11e7-bb16-000000000000 | read_mutation_data handling is done, sending a response to /127.0.0.1 | 106811168079939 | 127.0.0.4 |            114 | 212992572257781 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59748b1e-14e3-11e7-b84f-000000000000 |                      read_mutation_data: got response from /127.0.0.4 | 106811168079939 | 127.0.0.1 |           4189 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59748eb0-14e3-11e7-b84f-000000000000 |                                  read_mutation_data: querying locally | 106811168079939 | 127.0.0.1 |           4280 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59749a66-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.2 | 106811168079939 | 127.0.0.1 |           4580 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5974a59a-14e3-11e7-8cdc-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.2 |             12 | 269855355221015 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5974b1aa-14e3-11e7-8cdc-000000000000 | read_mutation_data handling is done, sending a response to /127.0.0.1 | 106811168079939 | 127.0.0.2 |            321 | 269855355221015 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5974dd46-14e3-11e7-b84f-000000000000 |                      read_mutation_data: got response from /127.0.0.2 | 106811168079939 | 127.0.0.1 |           6292 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5974e620-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.3 | 106811168079939 | 127.0.0.1 |           6519 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59770eef-14e3-11e7-88ef-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.3 |             14 | 138709993599978 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59771854-14e3-11e7-88ef-000000000000 | read_mutation_data handling is done, sending a response to /127.0.0.1 | 106811168079939 | 127.0.0.3 |            254 | 138709993599978 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59771c9e-14e3-11e7-b84f-000000000000 |                      read_mutation_data: got response from /127.0.0.3 | 106811168079939 | 127.0.0.1 |          21021 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59771e78-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.2 | 106811168079939 | 127.0.0.1 |          21068 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597721e6-14e3-11e7-8cdc-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.2 |              8 |  21115370758212 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5977256a-14e3-11e7-8cdc-000000000000 | read_mutation_data handling is done, sending a response to /127.0.0.1 | 106811168079939 | 127.0.0.2 |             98 |  21115370758212 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597727c3-14e3-11e7-b84f-000000000000 |                      read_mutation_data: got response from /127.0.0.2 | 106811168079939 | 127.0.0.1 |          21306 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59772892-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.4 | 106811168079939 | 127.0.0.1 |          21327 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59772bd5-14e3-11e7-bb16-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.4 |             10 | 238371111585204 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59772f45-14e3-11e7-bb16-000000000000 | read_mutation_data handling is done, sending a response to /127.0.0.1 | 106811168079939 | 127.0.0.4 |             98 | 238371111585204 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5977417f-14e3-11e7-b84f-000000000000 |                      read_mutation_data: got response from /127.0.0.4 | 106811168079939 | 127.0.0.1 |          21965 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59774262-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.5 | 106811168079939 | 127.0.0.1 |          21988 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59774482-14e3-11e7-a72a-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.5 |              8 | 199401997584972 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597746d1-14e3-11e7-a72a-000000000000 | read_mutation_data handling is done, sending a response to /127.0.0.1 | 106811168079939 | 127.0.0.5 |             67 | 199401997584972 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597748fc-14e3-11e7-b84f-000000000000 |                      read_mutation_data: got response from /127.0.0.5 | 106811168079939 | 127.0.0.1 |          22157 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597749a1-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.3 | 106811168079939 | 127.0.0.1 |          22173 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5977606c-14e3-11e7-88ef-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.3 |             12 | 248080828613419 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597768eb-14e3-11e7-88ef-000000000000 | read_mutation_data handling is done, sending a response to /127.0.0.1 | 106811168079939 | 127.0.0.3 |            229 | 248080828613419 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59777070-14e3-11e7-b84f-000000000000 |                      read_mutation_data: got response from /127.0.0.3 | 106811168079939 | 127.0.0.1 |          23167 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597771a5-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.4 | 106811168079939 | 127.0.0.1 |          23198 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5977761e-14e3-11e7-bb16-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.4 |             13 |  50620303064355 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597779f7-14e3-11e7-bb16-000000000000 | read_mutation_data handling is done, sending a response to /127.0.0.1 | 106811168079939 | 127.0.0.4 |            111 |  50620303064355 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59778a67-14e3-11e7-b84f-000000000000 |                      read_mutation_data: got response from /127.0.0.4 | 106811168079939 | 127.0.0.1 |          23831 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 59779434-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.3 | 106811168079939 | 127.0.0.1 |          24082 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597796cd-14e3-11e7-88ef-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.3 |              9 | 231099620585535 | shard 0
...
```
As you may see the query consists of repeating Coordinator-Replica sessions which take 4 traces to complete, e.g.:
```
 5973c320-14e3-11e7-b84f-000000000000 | 59742b71-14e3-11e7-b84f-000000000000 |                   read_mutation_data: sending a message to /127.0.0.5 | 106811168079939 | 127.0.0.1 |           1740 | 106811168079939 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597431f9-14e3-11e7-a72a-000000000000 |                  read_mutation_data: message received from /127.0.0.1 | 106811168079939 | 127.0.0.5 |             12 |  47600206259384 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 597435d8-14e3-11e7-a72a-000000000000 | read_mutation_data handling is done, sending a response to /127.0.0.1 | 106811168079939 | 127.0.0.5 |            111 |  47600206259384 | shard 0
 5973c320-14e3-11e7-b84f-000000000000 | 5974403c-14e3-11e7-b84f-000000000000 |                      read_mutation_data: got response from /127.0.0.5 | 106811168079939 | 127.0.0.1 |           2272 | 106811168079939 | shard 0
 ```

And we have  5192 traces in total, which gives us about 1300 such iterations.
This definitely seems too many.

**Comment by vladzcloudius** [comment]
Enabling `trace` info level I see a lot of these messages during the query:
```
TRACE 2017-03-29 21:12:32,104 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.1}
TRACE 2017-03-29 21:12:32,104 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.2}
TRACE 2017-03-29 21:12:32,104 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.1}
TRACE 2017-03-29 21:12:32,104 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,105 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.5}
TRACE 2017-03-29 21:12:32,105 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.1}
TRACE 2017-03-29 21:12:32,105 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.4}
TRACE 2017-03-29 21:12:32,105 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,105 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.3}
TRACE 2017-03-29 21:12:32,106 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,106 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.3}
TRACE 2017-03-29 21:12:32,107 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.2}
TRACE 2017-03-29 21:12:32,107 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,107 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.3}
TRACE 2017-03-29 21:12:32,107 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,108 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.4}
TRACE 2017-03-29 21:12:32,108 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.2}
TRACE 2017-03-29 21:12:32,108 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.5}
TRACE 2017-03-29 21:12:32,109 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.3}
TRACE 2017-03-29 21:12:32,109 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.1}
TRACE 2017-03-29 21:12:32,109 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.4}
TRACE 2017-03-29 21:12:32,109 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.5}
TRACE 2017-03-29 21:12:32,109 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.3}
TRACE 2017-03-29 21:12:32,110 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,110 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.5}
TRACE 2017-03-29 21:12:32,110 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,111 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.4}
TRACE 2017-03-29 21:12:32,111 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,111 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.5}
TRACE 2017-03-29 21:12:32,111 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.3}
TRACE 2017-03-29 21:12:32,112 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.5}
TRACE 2017-03-29 21:12:32,112 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.3}
TRACE 2017-03-29 21:12:32,112 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.5}
TRACE 2017-03-29 21:12:32,112 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.2}
TRACE 2017-03-29 21:12:32,112 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.4}
TRACE 2017-03-29 21:12:32,113 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,113 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.1}
TRACE 2017-03-29 21:12:32,113 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.4}
TRACE 2017-03-29 21:12:32,114 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,114 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.3}
TRACE 2017-03-29 21:12:32,114 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,115 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.1}
TRACE 2017-03-29 21:12:32,115 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.6}
TRACE 2017-03-29 21:12:32,115 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.5}
TRACE 2017-03-29 21:12:32,117 [shard 0] storage_proxy - creating range read executor with targets {127.0.0.1}
```
How many?
```
$ grep "creating range read executor with targets" ~/.ccm/scylla-6/node1/logs/system.log | wc -l
1314
```

**Comment by vladzcloudius** [comment]
@gleb-cloudius Any ideas?

**Comment by gleb-cloudius** [comment]
On Wed, Mar 29, 2017 at 06:26:20PM -0700, vladzcloudius wrote:
> @gleb-cloudius Any ideas?
> 
Nothing unexpected. You have 256*6 ranges. Current code, lacking
rows per range estimator, assumes that table is densely populated
and query one range at a time.

--
			Gleb.

**Comment by gleb-cloudius** [comment]
Can be closed as dup of #1230.

--
			Gleb.
